### PR TITLE
openjdk21, openjfx21: init at 21

### PR DESCRIPTION
**Pull request by Infinidoge**
## Description of changes

Updates the various Java packages to JDK 21, as per the release: https://openjdk.org/projects/jdk/21/

Temurin has not yet released official binaries for JDK 21 yet, as per [their blog post](https://adoptium.net/blog/2023/09/temurin21-delay/), and as such `generate-sources.py` does not pick up any versions to put in the corresponding JSON file.

Right now I have it setup to bootstrap `openjdk21` using `temurin-bin-20` instead.

I will update the commit, and remove the temporary commit, when Temurin has a release for JDK 21.

Please let me know if there is anything else I need to do.

Also of note, I removed the patches `fix-java-home-jdk10` in Nixpkgs, and `FixNullPtrCast` from Alpine, as they did not apply when building JDK 21. Please let me know if the patch needs to be modified, or if they can just be left out.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by AurelienRichez — COMMENTED** [review]
Glad someone started working on this. I was actually considering opening a PR on nixpkgs so I might as well share my findings here if that can help. I could try it out on macos intel (but not M1/M2).

About the patches, both bit of code have been modified:

- fix-java-home-jdk10: https://github.com/openjdk/jdk21u/blob/b12961d94d71c2a06dc1e44ae9c46aaec817e863/src/hotspot/os/linux/os_linux.cpp#L2610-L2617 . The `NULL` was replace by `nullptr`. So we could create a new patch but it's not clear to me what's is the original purpose and if it's still needed today.
- FixNullPtrCast: https://github.com/openjdk/jdk21u/blob/b12961d94d71c2a06dc1e44ae9c46aaec817e863/src/hotspot/cpu/aarch64/interpreterRT_aarch64.cpp#L270: looks like the change from the patch was applied in upstream so no need for that one anymore

File: pkgs/development/compilers/temurin-bin/jdk-linux.nix [hunk 1]
```diff
@@ -23,4 +23,7 @@ in
 
   jdk-20 = common { sourcePerArch = sources.jdk.openjdk20; };
   jre-20 = common { sourcePerArch = sources.jre.openjdk20; };
+
+  jdk-21 = common { sourcePerArch = sources.jdk.openjdk21; };
+  jre-21 = common { sourcePerArch = sources.jre.openjdk21; };
```
[the remark this answers is not in the document]
Comment (Infinidoge) [reply pkgs/development/compilers/temurin-bin/jdk-linux.nix:28]: Oh huh, didn't know that. I could leave out temurin for now and use temurin 20 to bootstrap JDK 21, then make a separate PR later for temurin 21.

**Comment by Infinidoge** [comment]
@wegank What did you change(?)

**Comment by wegank** [comment]
Nothing significant, just a rebase. This should not solve the ofborg eval error.

**Comment by Infinidoge** [comment]
I believe I can solve the OfBorg eval error by splitting the PR, if that would be preferable

**Comment by Infinidoge** [comment]
Times I've compiled OpenJDK 21: IIII

**Comment by Infinidoge** [comment]
Missed a spot when I was splitting the original commit, *now* it should eval :)

**Comment by wegank** [comment]
@ofborg build openjdk21 openjfx21

**Comment by reckenrode** [comment]
This PR broke nixpkgs-review on Darwin because it tries and fails to import `darwin/meta.nix`.

**Comment by wegank** [comment]
> This PR broke nixpkgs-review on Darwin because it tries and fails to import `darwin/meta.nix`.

I've just merged #259825, whose last commit should have solved this. Would you please test again?

**Comment by reckenrode** [comment]
> > This PR broke nixpkgs-review on Darwin because it tries and fails to import `darwin/meta.nix`.
> 
> I've just merged #259825, whose last commit should have solved this. Would you please test again?

Yep, it’s working now. Thanks! That was some “lucky” timing on my part. 😅

**Pull request merged**